### ISSUE TITLE
Unsupervised MNIST Image Clustering Accuracy

### ISSUE THREAD

**Issue by turetken** [xu-ji/IIC#36]
Thank you for sharing the source code for this great work!

I am trying to replicate your unsupervised image clustering results on the MNIST dataset. In the ArXiv paper, the avg. and lowest loss sub-head accuracies are reported to be 98.4% and 99.2% respectively. And in one of your answers to an issue, you stated that the loss function goes down to -2.20, which is also what I obtained after running the training for a few hundred epochs. However, I was never able to reach to a 98% or 99% accuracy. After training from scratch five times, the best accuracy I obtained was around 97.87% on the training and test datasets for all heads.

Do you know what I might have done wrong? Did you get the accuracies you reported on the paper with the same set of random transformations used in the training? 

Here is the command I used for training:

`export CUDA_VISIBLE_DEVICES=0 && python -m code.scripts.cluster.cluster_greyscale_twohead --model_ind 685 --arch ClusterNet6cTwoHead --mode IID --dataset MNIST --dataset_root /root/IIC/dataset/MNIST --out_root /root/IIC/results --gt_k 10 --output_k_A 50 --output_k_B 10  --lamb_A 1.0 --lamb_B 1.0 --lr 0.0001 --num_epochs 3200 --batch_sz 700 --num_dataloaders 5 --num_sub_heads 5 --crop_orig --crop_other --tf1_crop centre_half --tf2_crop random --tf1_crop_sz 20  --tf2_crop_szs 16 20 24 --input_sz 24 --rot_val 25 --no_flip --head_B_epochs 2`

And here is the result I got at the 606th epoch:

> Starting e_i: 606
Model ind 685 epoch 606 head B batch: 0 avg loss -2.212163 avg loss no lamb -2.212163 time 2019-12-13 09:59:16.075421
Model ind 685 epoch 606 head B batch: 100 avg loss -2.227755 avg loss no lamb -2.227755 time 2019-12-13 09:59:43.266961
Model ind 685 epoch 606 head B batch: 200 avg loss -2.232833 avg loss no lamb -2.232833 time 2019-12-13 10:00:10.573869
Model ind 685 epoch 606 head B batch: 300 avg loss -2.246190 avg loss no lamb -2.246190 time 2019-12-13 10:00:37.024877
Model ind 685 epoch 606 head B batch: 400 avg loss -2.200957 avg loss no lamb -2.200957 time 2019-12-13 10:01:02.510503
Model ind 685 epoch 606 head B batch: 0 avg loss -2.246450 avg loss no lamb -2.246450 time 2019-12-13 10:01:30.403862
Model ind 685 epoch 606 head B batch: 100 avg loss -2.209442 avg loss no lamb -2.209442 time 2019-12-13 10:01:57.819133
Model ind 685 epoch 606 head B batch: 200 avg loss -2.216221 avg loss no lamb -2.216221 time 2019-12-13 10:02:25.090045
Model ind 685 epoch 606 head B batch: 300 avg loss -2.232495 avg loss no lamb -2.232495 time 2019-12-13 10:02:52.324494
Model ind 685 epoch 606 head B batch: 400 avg loss -2.208003 avg loss no lamb -2.208003 time 2019-12-13 10:03:20.211105
Model ind 685 epoch 606 head A batch: 0 avg loss -2.237204 avg loss no lamb -2.237204 time 2019-12-13 10:03:47.252569
Model ind 685 epoch 606 head A batch: 100 avg loss -2.212928 avg loss no lamb -2.212928 time 2019-12-13 10:04:14.799001
Model ind 685 epoch 606 head A batch: 200 avg loss -2.218388 avg loss no lamb -2.218388 time 2019-12-13 10:04:42.189923
Model ind 685 epoch 606 head A batch: 300 avg loss -2.255437 avg loss no lamb -2.255437 time 2019-12-13 10:05:09.242370
Model ind 685 epoch 606 head A batch: 400 avg loss -2.210399 avg loss no lamb -2.210399 time 2019-12-13 10:05:36.609692
Pre: time 2019-12-13 10:06:12.301906:
        std: 2.8002994e-05
        best_train_sub_head_match: [(0, 7), (1, 9), (2, 0), (3, 8), (4, 3), (5, 4), (6, 6), (7, 5), (8, 2), (9, 1)]
        test_accs: [0.97877145, 0.97877145, 0.9787143, 0.9787143, 0.9787143]
        train_accs: [0.97877145, 0.97877145, 0.9787143, 0.9787143, 0.9787143]
        best_train_sub_head: 0
        worst: 0.9787143
        avg: 0.9787372
        best: 0.97877145

**Comment by xu-ji** [comment]
Yes, model 685 (and the transforms specified for it) are exactly what were used for the paper results, and if you download the model you can check the accuracy.

I do think training is, as always, stochastic and we got a little lucky with our initialisation. But 97.9% is already very close. It's only 1.3% less than 99.2%. A [reimplementation of IIC in tensorflow](https://github.com/astirn/IIC/blob/master/README.md#running-the-code) has also attained 99.2%. But performance on MNIST, as you can tell from our table 1, is already saturated and it's really our STL, CIFAR and segmentation results that are notable.

I would suggest to run it a few more times. You could also experiment more with the transforms used, for example salt and pepper noise might be quite useful and we didn't use that.
